// material-ui-form-builder/components/AlertDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Button = _interopRequireDefault(require("@material-ui/core/Button"));

var _Dialog = _interopRequireDefault(require("@material-ui/core/Dialog"));

var _DialogActions = _interopRequireDefault(require("@material-ui/core/DialogActions"));

var _DialogContent = _interopRequireDefault(require("@material-ui/core/DialogContent"));

var _DialogContentText = _interopRequireDefault(require("@material-ui/core/DialogContentText"));

var _DialogTitle = _interopRequireDefault(require("@material-ui/core/DialogTitle"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function AlertDialog(_ref) {
  var userChoice = _ref.userChoice,
      handleOpen = _ref.handleOpen,
      content = _ref.content,
      title = _ref.title;
  var defaultMessage = 'There is no provided message, please add one! ';
  return _react.default.createElement("div", null, _react.default.createElement(_Dialog.default, {
    open: handleOpen,
    "aria-labelledby": "alert-dialog-title",
    "aria-describedby": "alert-dialog-description"
  }, _react.default.createElement(_DialogTitle.default, {
    id: "alert-dialog-title"
  }, title ? title : null), _react.default.createElement(_DialogContent.default, null, _react.default.createElement(_DialogContentText.default, {
    id: "alert-dialog-description"
  }, content ? content : defaultMessage)), _react.default.createElement(_DialogActions.default, null, _react.default.createElement(_Button.default, {
    onClick: function onClick() {
      return userChoice('cancel');
    },
    color: "primary"
  }, "Cancel"), _react.default.createElement(_Button.default, {
    onClick: function onClick() {
      return userChoice('ok');
    },
    color: "primary",
    autoFocus: true
  }, "Ok"))));
}

var _default = AlertDialog;
exports.default = _default;
