// material-ui-form-builder/components/ChipInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = require("prop-types");

var _classnames = _interopRequireDefault(require("classnames"));

var _Creatable = _interopRequireDefault(require("react-select/lib/Creatable"));

var _styles = require("@material-ui/core/styles");

var _Typography = _interopRequireDefault(require("@material-ui/core/Typography"));

var _TextField = _interopRequireDefault(require("@material-ui/core/TextField"));

var _Chip = _interopRequireDefault(require("@material-ui/core/Chip"));

var _Cancel = _interopRequireDefault(require("@material-ui/icons/Cancel"));

var _colorManipulator = require("@material-ui/core/styles/colorManipulator");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var styles = function styles(theme) {
  return {
    root: {
      flexGrow: 1
    },
    input: {
      display: 'flex',
      padding: 0
    },
    valueContainer: {
      display: 'flex',
      flexWrap: 'wrap',
      flex: 1,
      alignItems: 'center',
      overflow: 'hidden'
    },
    chip: {
      margin: "2px 4px 2px 0"
    },
    chipFocused: {
      backgroundColor: (0, _colorManipulator.emphasize)( // theme.palette.type === 'light' ? theme.palette.grey[300] : theme.palette.grey[700],
      // theme.palette.grey[300],
      '#DCDCDC', 0.08)
    },
    placeholder: {
      position: 'absolute',
      left: 2,
      fontSize: 14
    }
  };
};

function inputComponent(_ref) {
  var inputRef = _ref.inputRef,
      props = _objectWithoutProperties(_ref, ["inputRef"]);

  return _react.default.createElement("div", _extends({
    ref: inputRef
  }, props));
}

function Control(props) {
  return _react.default.createElement(_TextField.default, _extends({
    fullWidth: true,
    InputProps: {
      inputComponent: inputComponent,
      inputProps: _objectSpread({
        className: props.selectProps.classes.input,
        inputRef: props.innerRef,
        children: props.children
      }, props.innerProps)
    }
  }, props.selectProps.textFieldProps));
}

function Placeholder(props) {
  return _react.default.createElement(_Typography.default, _extends({
    color: "textSecondary",
    className: props.selectProps.classes.placeholder
  }, props.innerProps), props.children);
}

function ValueContainer(props) {
  return _react.default.createElement("div", {
    className: props.selectProps.classes.valueContainer
  }, props.children);
}

function MultiValue(props) {
  return _react.default.createElement(_Chip.default, {
    tabIndex: -1,
    label: props.children,
    className: (0, _classnames.default)(props.selectProps.classes.chip, _defineProperty({}, props.selectProps.classes.chipFocused, props.isFocused)),
    onDelete: props.removeProps.onClick,
    deleteIcon: _react.default.createElement(_Cancel.default, props.removeProps)
  });
}

var getName = function getName(name) {
  return _lodash.default.replace(name, /\./g, ' ');
};

var components = {
  Control: Control,
  MultiValue: MultiValue,
  Placeholder: Placeholder,
  ValueContainer: ValueContainer,
  DropdownIndicator: null
};

var ChipInput = function ChipInput(_ref2) {
  var field = _ref2.field,
      value = _ref2.value,
      _onChange = _ref2.onChange,
      errors = _ref2.errors,
      classes = _ref2.classes;

  var _React$useState = _react.default.useState(''),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      inputValue = _React$useState2[0],
      setInputValue = _React$useState2[1];

  var key = field.key || field.name || 'key';
  var name = field.name || 'field';

  var errorText = _lodash.default.get(errors, key, false);

  var handleInputChange = function handleInputChange(inputValue) {
    setInputValue(inputValue);
  };

  var handleKeyDown = function handleKeyDown(event) {
    if (!inputValue) return;

    switch (event.key) {
      case 'Enter':
      case 'Tab':
        _onChange([].concat(_toConsumableArray(value), [inputValue]));

        setInputValue('');
        event.preventDefault();
    }
  };

  return _react.default.createElement("div", null, _react.default.createElement(_Creatable.default, {
    classes: classes,
    components: components,
    inputValue: inputValue,
    value: value.map(function (chip) {
      return {
        label: chip,
        value: chip
      };
    }),
    onChange: function onChange(items) {
      _onChange(_lodash.default.map(items, function (item) {
        return item.value;
      }));
    },
    textFieldProps: {
      label: getName(name),
      InputLabelProps: {
        shrink: true
      }
    },
    onInputChange: handleInputChange,
    onKeyDown: handleKeyDown,
    isClearable: true,
    isMulti: true,
    menuIsOpen: false
  }), errorText && _react.default.createElement("span", {
    style: {
      fontSize: '10px',
      color: 'red',
      fontFamily: '"Roboto", "Helvetica", "Arial", sans-serif'
    }
  }, errorText));
};

ChipInput.defaultProps = {
  errors: {}
};
ChipInput.propTypes = {
  field: (0, _propTypes.shape)({
    name: _propTypes.string,
    type: _propTypes.string,
    key: _propTypes.string
  }).isRequired,
  value: (0, _propTypes.arrayOf)((0, _propTypes.oneOfType)([_propTypes.string, _propTypes.number])).isRequired,
  onChange: _propTypes.func.isRequired,
  errors: (0, _propTypes.shape)({})
};

var _default = (0, _styles.withStyles)(styles)(ChipInput);

exports.default = _default;
